src/display/node_stream.js

```diff
@@ -0,0 +1,351 @@
+  constructor(options) {
+    this.options = options;
+    this.source = options.source;
+    this.url = url.parse(this.source.url);
+    this.isHttp = this.url.protocol === 'http:' ||
+                  this.url.protocol === 'https:';
+    this.isFsUrl = this.url.protocol === 'file:' || !this.url.host;
+    this.httpHeaders = (this.isHttp && this.source.httpHeaders) || {};
+
+    this._fullRequest = null;
+    this._rangeRequestReaders = [];
+  }
+
+  getFullReader() {
+    assert(!this._fullRequest);
+    this._fullRequest = this.isFsUrl ?
+      new PDFNodeStreamFsFullReader(this) :
+      new PDFNodeStreamFullReader(this);
+    return this._fullRequest;
+  }
+
+  getRangeReader(start, end) {
+    let rangeReader = this.isFsUrl ?
+      new PDFNodeStreamFsRangeReader(this, start, end) :
+      new PDFNodeStreamRangeReader(this, start, end);
+    this._rangeRequestReaders.push(rangeReader);
+    return rangeReader;
+  }
+
+  cancelAllRequests(reason) {
+    if (this._fullRequest) {
+      this._fullRequest.cancel(reason);
+    }
+
+    let readers = this._rangeRequestReaders.slice(0);
+    readers.forEach(function(reader) {
+      reader.cancel(reason);
+    });
+  }
+}
+
+class BaseFullReader {
+  constructor(stream) {
+    this._url = stream.url;
+    this._done = false;
+    this._errored = false;
+    this._reason = null;
+    this.onProgress = null;
+    this._length = stream.source.length;
+    this._loaded = 0;
+
+    this._disableRange = stream.options.disableRange || false;
+    this._rangeChunkSize = stream.source.rangeChunkSize;
+    if (!this._rangeChunkSize && !this._disableRange) {
+      this._disableRange = true;
+    }
+
+    this._isStreamingSupported = !stream.source.disableStream;
+    this._isRangeSupported = !stream.options.disableRange;
+
+    this._readableStream = null;
+    this._readCapability = createPromiseCapability();
+    this._headersCapability = createPromiseCapability();
+  }
+
+  get headersReady() {
+    return this._headersCapability.promise;
+  }
+
+  get contentLength() {
+    return this._length;
+  }
+
+  get isRangeSupported() {
+    return this._isRangeSupported;
+  }
+
+  get isStreamingSupported() {
+    return this._isStreamingSupported;
+  }
+
+  read() {
+    return this._readCapability.promise.then(() => {
+      if (this._done) {
+        return Promise.resolve({ value: undefined, done: true, });
+      }
+      if (this._errored) {
+        return Promise.reject(this._reason);
+      }
+
+      let chunk = this._readableStream.read();
+      if (chunk === null) {
+        this._readCapability = createPromiseCapability();
+        return this.read();
+      }
+      this._loaded += chunk.length;
+      if (this.onProgress) {
+        this.onProgress({
+          loaded: this._loaded,
+          total: this._length,
+        });
+      }
+      return Promise.resolve({ value: chunk, done: false, });
+    });
+  }
+
+  cancel(reason) {
+    this._readableStream.destroy(reason);
```

Comment:
here and at the range base reader, we need to think "what if cancel is called before _readableStream is set". I recommend to extract `on('error'` body into _error(reason) method and call it when this._readableStream is not set. in _setReadableStream we can check if `this._errored = true` we can quickly destroy it.